components/proxy/src/test/kotlin/com/hotels/styx/routing/handlers/HostProxyTest.kt

```diff
@@ -0,0 +1,131 @@
+/*
+  Copyright (C) 2013-2019 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.routing.handlers
+
+import com.google.common.net.HostAndPort
+import com.hotels.styx.api.Eventual
+import com.hotels.styx.api.HttpRequest
+import com.hotels.styx.api.HttpResponse
+import com.hotels.styx.api.HttpResponseStatus.OK
+import com.hotels.styx.api.LiveHttpRequest
+import com.hotels.styx.client.StyxHostHttpClient
+import com.hotels.styx.routing.RoutingContext
+import com.hotels.styx.routing.handle
+import com.hotels.styx.routing.routingObjectDef
+import io.kotlintest.TestCase
+import io.kotlintest.shouldBe
+import io.kotlintest.shouldThrow
+import io.kotlintest.specs.FeatureSpec
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.verify
+import reactor.core.publisher.toMono
+
+class HostProxyTest : FeatureSpec() {
+    val request = HttpRequest.get("/").build()
+    var client: StyxHostHttpClient? = null
+
+    init {
+        feature("Routing and proxying") {
+            scenario("Proxies traffic") {
+                HostProxy(HostAndPort.fromString("localhost:80"), client).let {
+                    it.handle(request.stream(), mockk())
+                }
+
+                verify {
+                    client?.sendRequest(ofType(LiveHttpRequest::class))
+                }
+            }
+
+            scenario("Requests arriving at stopped HostProxy object") {
+                val exception = HostProxy(HostAndPort.fromString("localhost:80"), client).let {
+                    it.stop()
```

Comment:
Should our test know that `stop` is synchronous, despite returning a `CompleteableFuture`? If it becomes async at any point in the future this test will start failing.

Reply:
Not at the moment. We'll fix this when it starts failing.

I have a separate work item in the pipeline to fix it. But that goes beyond scope of this PR.

Reply:
I think my comment was unclear. What I was saying is that instead of calling `stop()` here, we should call something like `stop().get()`, since we should assume that a method with an asynchronous signature will execute asynchronously, even if we know that this particular case doesn't.

Reply:
Actually no. The handler should start rejecting requests right after a call to `stop`, before the future has completed, and the handler has fully come to stop. The existing requests ought to drain gracefully. But that starts going out of scope and therefore is a separate enhancement.